staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go

```diff
@@ -301,25 +310,50 @@ type jsonPatcher struct {
 	fieldManager *fieldmanager.FieldManager
 }
 
-func (p *jsonPatcher) applyPatchToCurrentObject(currentObject runtime.Object) (runtime.Object, error) {
+func (p *jsonPatcher) applyPatchToCurrentObject(requestContext context.Context, currentObject runtime.Object) (runtime.Object, error) {
 	// Encode will convert & return a versioned object in JSON.
 	currentObjJS, err := runtime.Encode(p.codec, currentObject)
 	if err != nil {
 		return nil, err
 	}
 
 	// Apply the patch.
-	patchedObjJS, err := p.applyJSPatch(currentObjJS)
+	patchedObjJS, appliedStrictErrs, err := p.applyJSPatch(currentObjJS)
 	if err != nil {
 		return nil, err
 	}
 
 	// Construct the resulting typed, unversioned object.
 	objToUpdate := p.restPatcher.New()
 	if err := runtime.DecodeInto(p.codec, patchedObjJS, objToUpdate); err != nil {
-		return nil, errors.NewInvalid(schema.GroupKind{}, "", field.ErrorList{
-			field.Invalid(field.NewPath("patch"), string(patchedObjJS), err.Error()),
-		})
+		strictError, isStrictError := runtime.AsStrictDecodingError(err)
+		switch {
+		case !isStrictError:
+			// disregard any appliedStrictErrs, because it's an incomplete
+			// list of strict errors given that we don't know what fields were
+			// unknown because DecodeInto failed. Non-strict errors trump in this case.
+			return nil, errors.NewInvalid(schema.GroupKind{}, "", field.ErrorList{
+				field.Invalid(field.NewPath("patch"), string(patchedObjJS), err.Error()),
+			})
+		case p.validationDirective == metav1.FieldValidationStrict:
+			strictDecodingError := runtime.NewStrictDecodingError(append(appliedStrictErrs, strictError.Errors()...))
+			return nil, errors.NewInvalid(schema.GroupKind{}, "", field.ErrorList{
+				field.Invalid(field.NewPath("patch"), string(patchedObjJS), strictDecodingError.Error()),
+			})
+		case p.validationDirective == metav1.FieldValidationWarn:
+			strictDecodingError, _ := runtime.AsStrictDecodingError(runtime.NewStrictDecodingError(append(appliedStrictErrs, strictError.Errors()...)))
+			addStrictDecodingWarnings(requestContext, strictDecodingError.Errors())
```

Comment:
simplify to pass errors array directly to addStrictDecodingWarnings instead of wrapping into NewStrictDecodingError and then unwrapping